cpp/src/table/row_operators.cu

```diff
@@ -373,51 +383,467 @@ namespace row {
+
+        return {std::move(transformed_lhs),
+                std::optional<column_view>{std::move(transformed_rhs)},
+                std::move(out_cols_lhs),
+                std::move(out_cols_rhs)};
+      } else {  // rhs table is not available
+        out_cols_lhs.emplace_back(compute_ranks(child_lhs));
+        auto transformed_lhs = replace_child(lhs, out_cols_lhs.back()->view(), out_cols_lhs);
+
+        return {std::move(transformed_lhs),
+                std::nullopt,
+                std::move(out_cols_lhs),
+                std::move(out_cols_rhs)};
+      }
+    }
+    // Found a lists-of-lists column.
+    else if (child_lhs.type().id() == type_id::LIST) {
+      auto const child_rhs_opt =
+        rhs_opt
+          ? std::optional<column_view>{cudf::lists_column_view{rhs_opt.value()}.get_sliced_child(
+              stream)}
+          : std::nullopt;
+
+      // Recursively call transformation on the child column.
+      auto [new_child_lhs, new_child_rhs_opt, out_cols_child_lhs, out_cols_child_rhs] =
+        transform_lists_of_structs(child_lhs, child_rhs_opt, column_null_order, stream);
+
+      // Only transform the current pair of columns if their children have been transformed.
+      if (out_cols_child_lhs.size() > 0 || out_cols_child_rhs.size() > 0) {
+        out_cols_lhs.insert(out_cols_lhs.end(),
+                            std::make_move_iterator(out_cols_child_lhs.begin()),
+                            std::make_move_iterator(out_cols_child_lhs.end()));
+        out_cols_rhs.insert(out_cols_rhs.end(),
+                            std::make_move_iterator(out_cols_child_rhs.begin()),
+                            std::make_move_iterator(out_cols_child_rhs.end()));
+
+        auto transformed_lhs = replace_child(lhs, new_child_lhs, out_cols_lhs);
+        if (rhs_opt) {
+          auto transformed_rhs =
+            replace_child(rhs_opt.value(), new_child_rhs_opt.value(), out_cols_rhs);
+
+          return {std::move(transformed_lhs),
+                  std::optional<column_view>{std::move(transformed_rhs)},
+                  std::move(out_cols_lhs),
+                  std::move(out_cols_rhs)};
+        } else {
+          return {std::move(transformed_lhs),
+                  std::nullopt,
+                  std::move(out_cols_lhs),
+                  std::move(out_cols_rhs)};
+        }
+      }
+    }
+    // else == child is not STRUCT or LIST: just go to the end of this function, no transformation.
+  }
+  // Any structs-of-lists should be decomposed into empty struct type `Struct<>` before being
+  // processed by this function.
+  else if (lhs.type().id() == type_id::STRUCT) {
+    CUDF_EXPECTS(std::all_of(lhs.child_begin(),
+                             lhs.child_end(),
+                             [](auto const& child) { return child.type().id() != type_id::LIST; }),
+                 "Structs columns should be decomposed before reaching this function.");
```

Comment:
```suggestion
                 "Structs columns with list column descendants should be decomposed before reaching this function.");
```

Reply:
The message is correct: `decompose_struct` should be called before reaching this function, regardless of having lists column descendants or not.

Reply:
The order can't be checked here, we can only detect wrong order when there is structs-of-lists column exists.

Reply:
> regardless of having lists column descendants or not.
But isn't the check just performing if a struct column has a list column as a direct child? This error won't be thrown if it's just a struct column

Reply:
I see your point that after decomposition of `Struct<List, float>` we will have `Struct<>, List, float`. I think we need to phrase the comment better to explain why we only check for `Struct<List>`. What if there's `Struct<Struct<List>>` and it wasn't decomposed before calling this function?

Reply:
That's what I said above:
> The order can't be checked here, we can only detect wrong order when there is structs-of-lists column exists.

Reply:
> `Struct<Struct<List>>`

That will be decomposed into a  table of `[Struct<Struct<>>, Struct<>, List<>]` and it will pass the check.